spec/tbd_tests_spec.rb

```diff
@@ -1566,3 +1566,414 @@
+              "title": "Unique PSI set identifier",
+              "type": "string"
+            },
+            "rimjoist": {
+              "title": "Floor/wall edge PSI",
+              "type": "number",
+              "minimum": 0
+            },
+            "parapet": {
+              "title": "Roof/wall or exposed-floor/wall edge PSI",
+              "type": "number",
+              "minimum": 0
+            },
+            "fenestration": {
+              "title": "Window, door, skylight perimeter PSI",
+              "type": "number",
+              "minimum": 0
+            },
+            "concave": {
+              "title": "Wall corner [0°,135°) PSI",
+              "type": "number",
+              "minimum": 0
+            },
+            "convex": {
+              "title": "Wall corner (225°,360°] PSI",
+              "type": "number",
+              "minimum": 0
+            },
+            "balcony": {
+              "title": "Floor/balcony edge PSI ",
+              "type": "number",
+              "minimum": 0
+            },
+            "party": {
+              "title": "Party wall edge PSI",
+              "type": "number",
+              "minimum": 0
+            },
+            "grade": {
+              "title": "Floor/foundation edge PSI",
+              "type": "number",
+              "minimum": 0
+            }
+          },
+          "additionalProperties": false,
+          "required": [
+            "id"
+          ],
+          "minProperties": 2
+        },
+        "KHI": {
+          "description": "KHI-value (in W/K) for a point thermal bridge",
+          "type": "object",
+          "properties": {
+            "id": {
+              "title": "Unique KHI identifier",
+              "type": "string"
+            },
+            "point": {
+              "title": "Point KHI-value",
+              "type": "number",
+              "minimum": 0
+            }
+          },
+          "additionalProperties": false,
+          "required": [
+            "id",
+            "point"
+          ]
+        },
+        "Edge": {
+          "description": "Surface(s) edge as thermal bridge",
+          "type": "object",
+          "properties": {
+            "psi": {
+              "title": "PSI-set identifier",
+              "type": "string"
+            },
+            "type": {
+              "title": "PSI-set type, e.g. 'parapet'",
+              "type": "string"
+            },
+            "length": {
+              "title": "Edge length (m), 10cm min",
+              "type": "number",
+              "minimum": 0.1
+            },
+            "surfaces": {
+              "title": "Surface(s) connected to edge",
+              "type": "array",
+              "items": {
+                "type": "string"
+              },
+              "minItems": 1,
+              "uniqueItems": true
+            }
+          },
+          "additionalProperties": false,
+          "required": [
+            "psi",
```

Comment:
If psi is required for each edge, how do you specify that an edge uses a default value?

Reply:
Do you just not list edges that are defaulted?

Reply:
In 99% of cases, one shouldn't see "edges" in a TBD JSON file (when used as input ... for output, that's a whole other thing).

The "edges" input option is offered for (i) the odd cases where a misdiagnosis occurs (i.e. TBD misidentifies the edge type) or (ii) when an edge isn't in the OSM file. These are fairly rare cases.

Reply:
Although my initial reply correctly describes the context where one would expect to encounter custom "edges" in a TBD JSON file (as input), I just noticed I wasn't directly answering your 2 questions. Here goes:

**1.** In circumstances where a TBD JSON file is used as input, there are **2** cases (**a** & **b**) to consider for "edges":

**a.** edges that are **absent** from the OSM geometry (and hence absent from the Topolys model): Impossible for TBD to diagnose the thermal bridge (TB) type (e.g. "rimjoist" ? "balcony" ?) on its own - users must inform TBD of the TB type (e.g. "balcony"). That being said, I see your point in this specific instance: users should nonetheless have the option of _defaulting_ the edge PSI-value to whatever "Surface", "Space", "Story" or building "Unit" default set would apply. I have since removed the "psi" key from the _required_ list of _properties_ for "Edges". (Good call, thanks). Upcoming commit ...

**b.** edges, although **present** in the Topolys model, that users nonetheless want to (i) override the TBD-identified TB type (e.g. "balcony" rather than "rimjoist"), or (ii) override the assigned PSI-value (superseding whatever default PSI-set would apply). Situation (i) is similar to case (a) above: users should also have the option of _defaulting_ the PSI-value. However in situation (ii), one would expect users to specify the PSI-value/set ... as this is the likely objective.

**2a.** "Do you just not list edges that are defaulted?" ... Based on the previous answer, one can now list any edge (present or absent from the Topolys model), while _defaulting_ (or not) its PSI-value.

**2b.** "How would you know the type of the defaulted edges?" ... in cases 1a & 1b(i) above, this is likely what a user would actually want, i.e. to override an edge's TBD-assigned type (e.g. "balcony" instead of "rimjoist"). This is analogous to resetting an OpenStudio surface type (e.g. a slanted "floor", instead of an auto-assigned "wall"). For case 1b(ii), that's simply how one (re)sets the PSI-value of an edge: PSI-set + TB type. It would be odd if a user could pinpoint an individual edge to customize and not know it's nature ...

In light of this, I have also decided to remove "length" as a _required_ property. A list of 2x surfaces (that share the edge) should be enough to pinpoint the edge in the Topolys model. However, as more than one edge can be shared between 2x surfaces, a provided "length" would narrow down the search. For cases 1a, the "length" is absolutely required - so will be looking more at post-schema validation in that case. Thanks for raising these questions.